tutorials/securing-your-app-with-spring-security/04__form-less/content.adoc

```diff
@@ -0,0 +1,141 @@
+
+            // Allow all requests by logged in users.
+            .anyRequest().authenticated()
+
+            // Configure the login page.
+            .and().formLogin().loginPage("/" + LoginView.ROUTE).permitAll() // <3>
+
+            // Configure logout
+            .and().logout().logoutSuccessUrl(LOGOUT_SUCCESS_URL);
+}
+----
+<1> Since SpringBoot 2 the authentication manager bean has to be exposed manually.
+<2> We also have to expose the custom request cache to access it in our login view.
+<3> We keep the form login semi-activate even if we do not use it directly. This makes Spring Security take care of the redirects.
+
+The custom request cache needs some extensions to easily extract the redirect URL. It extends the `HttpSessionRequestCache` that internally stores the saved request as a `DefaultSavedRequest` in the session. That is why it is save to cast the `SavedRequest`.
+
+.`*CustomRequestCache.java*`
+[source,java,linenums]
+----
+public String resolveRedirectUrl() {
+	SavedRequest savedRequest = getRequest(VaadinServletRequest.getCurrent().getHttpServletRequest(), VaadinServletResponse.getCurrent().getHttpServletResponse());
+	if(savedRequest instanceof DefaultSavedRequest) {
+		final String requestURI = ((DefaultSavedRequest) savedRequest).getRequestURI(); // <1>
+		// check for valid URI and prevent redirecting to the login view
+		if (requestURI != null && !requestURI.isEmpty() && !requestURI.contains(LoginView.ROUTE)) { // <2>
+			return requestURI.startsWith("/") ? requestURI.substring(1) : requestURI; // <3>
+		}
+	}
+
+	// if everything fails, redirect to the main view
+	return "";
+}
+----
+<1> Here we use our custom request cache to get the saved redirect URL.
+<2> As the request is only saved in exceptional cases (user is not logged in) by Spring Security, we make sure we do not run into a NPE here. Additionally, we prevent redirects to the login view.
+<3> Some mangling to satisfy `Ui.navigate()` that expects relative links without a leading slash.
+
+Finally, we will put everything together in the login view itself:
+
+.`*LoginView.java*`
+[source,java,linenums]
+----
+@Tag("sa-login-view")
+@Route(value = LoginView.ROUTE)
+@PageTitle("Login")
+public class LoginView extends VerticalLayout {
+    public static final String ROUTE = "login";
+
+    private LoginOverlay login = new LoginOverlay(); // <1>
+
+    @Autowired
+    public LoginView(AuthenticationManager authenticationManager, // <2>
+                     CustomRequestCache requestCache) {
+        // configures login dialog and adds it to the main view
+        login.setOpened(true);
+        login.setTitle("Spring Secured Vaadin");
+        login.setDescription("Login Overlay Example");
+
+        add(login);
+
+        login.addLoginListener(e -> { // <3>
+            try {
+                // try to authenticate with given credentials, should always return !null or throw an {@link AuthenticationException}
```

Comment:
!null isn't legal Java AFAIK, but I know what you mean... might read easier as "not null"?